pkg/internal/source/kind.go

```diff
@@ -4,47 +4,55 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"reflect"
 	"time"
 
 	"k8s.io/apimachinery/pkg/api/meta"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/util/workqueue"
+
 	"sigs.k8s.io/controller-runtime/pkg/cache"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/handler"
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
 )
 
 // Kind is used to provide a source of events originating inside the cluster from Watches (e.g. Pod Create).
-type Kind struct {
+type Kind[T client.Object] struct {
```

Comment:
Technically this can also be any, if the `Kind` will be able to accept GVK and use `GetInfromerForKind`. I suppose users will have to write their own `[any]` source for now.

Reply:
> Technically this can also be any, if the Kind will be able to accept GVK and use GetInfromerForKind. I suppose users will have to write their own [any] source for now.

Not sure I understand. The `Source` has to have a type in the golang sense. There is no way to express a GVK as a type in the golang sense other than using a `client.Object` from which we then derive the GVK.

Reply:
This is not what I mean, GVK can’t be used as a type, true. But if you are able to supply a GVK externally (a field in the struct) then the generic type supplied may be anything. GVK can be constructed manually, and the object Kind will be able to work with something which does not need to be anywhere close to `client.Object`. It’s for the case of “build arbitrary controllers that might not even act on kube resources” as I saw it from the issue discussion.

Reply:
> It’s for the case of “build arbitrary controllers that might not even act on kube resources” as I saw it from the issue discussion.

Right, but `source.Kind` isn't meant for that but specifically for kube resources.